lib/tbot/services/application/proxy_service.go

```diff
@@ -0,0 +1,267 @@
+			cfg:                       cfg,
+			identityGenerator:         deps.IdentityGenerator,
+			clientBuilder:             deps.ClientBuilder,
+		}
+
+		svc.log = deps.LoggerForService(svc)
+		svc.statusReporter = deps.StatusRegistry.AddService(svc.String())
+		return svc, nil
+	}
+}
+
+type ProxyService struct {
+	connCfg                   connection.Config
+	defaultCredentialLifetime bot.CredentialLifetime
+	cfg                       *ProxyServiceConfig
+	proxyPinger               connection.ProxyPinger
+	log                       *slog.Logger
+	botClient                 *apiclient.Client
+	getBotIdentity            func() *identity.Identity
+	botIdentityReadyCh        <-chan struct{}
+	statusReporter            readyz.Reporter
+	identityGenerator         *identity.Generator
+	clientBuilder             *client.Builder
+
+	cache *utils.FnCache
+}
+
+func (s *ProxyService) Run(ctx context.Context) error {
+	ctx, span := applicationProxyTracer.Start(ctx, "ProxyService/Run")
+	defer span.End()
+
+	l := s.cfg.Listener
+
+	if l == nil {
+		s.log.DebugContext(ctx, "Opening listener for application tunnel", "listen", s.cfg.Listen)
+		var err error
+		l, err = internal.CreateListener(ctx, s.log, s.cfg.Listen)
+		if err != nil {
+			return trace.Wrap(err, "opening listener")
+		}
+		defer func() {
+			if err := l.Close(); err != nil && !utils.IsUseOfClosedNetworkError(err) {
+				s.log.ErrorContext(ctx, "Failed to close listener", "error", err)
+			}
+		}()
+	}
+
+	// Initialize the fnCache
+	fnCache, _ := utils.NewFnCache(utils.FnCacheConfig{
+		TTL: 1 * time.Minute,
+	})
+	s.cache = fnCache
+
+	// lp.Start will block and continues to block until lp.Close() is called.
+	// Despite taking a context, it will not exit until the first connection is
+	// made after the context is canceled.
+	var errCh = make(chan error, 1)
+	go func() {
+		s.log.DebugContext(ctx, "Starting proxy goroutine")
+		errCh <- s.startProxy()
+	}()
+	s.log.InfoContext(ctx, "Listening for proxy connections.", "address", l.Addr().String())
+
+	s.statusReporter.Report(readyz.Healthy)
+
+	select {
+	case <-ctx.Done():
+		return nil
+	case err := <-errCh:
+		s.statusReporter.ReportReason(readyz.Unhealthy, err.Error())
+		return trace.Wrap(err, "local proxy failed")
+	}
+}
+
+func (s *ProxyService) String() string {
+	return cmp.Or(
+		s.cfg.Name,
+		fmt.Sprintf("%s:%s:%s", ProxyServiceType, s.cfg.Listen, s.cfg.Applications),
+	)
+}
+
+func (s *ProxyService) issueCert(
+	ctx context.Context,
+	appName string,
+) (*tls.Certificate, types.Application, error) {
+	ctx, span := applicationProxyTracer.Start(ctx, "ProxyService/issueCert")
+	defer span.End()
+
+	// Right now we have to redetermine the route to app each time as the
+	// session ID may need to change. Once v17 hits, this will be automagically
+	// calculated by the auth server on cert generation, and we can fetch the
+	// routeToApp once.
+	effectiveLifetime := cmp.Or(s.cfg.CredentialLifetime, s.defaultCredentialLifetime)
```

Comment:
It feels like it'd be beneficial to extract the generation of the impersonated identity/client into a separate background goroutine. There's no need for us to complete this step for each upstream app - we only need to generate the routedIdent on an app by app basis. This should significantly reduce the amount of resources/latency for each first connection to an upstream app - since creating a client and the connection it opens is fairly expensive and slow.